### DISCUSSION TITLE
Why should we avoid joins between fact tables? And what are some potential solutions for cross-process analysis?

### DISCUSSION THREAD

**Discussion by diegogalanhoffman** [dbt-labs/devhub-writing#55] [Guides and playbooks]
OVERVIEW 
"As a general rule, fact tables shouldn't be directly joined. Most fact tables have a 1:M or M:M relationship, which results in the facts being overstated when measures are calculated." - _Lawrence Corr with Jim Stagnitto_ **Agile Data Warehouse Design**

Is this general rule a timeless principal of dimensional modeling as first defined by Ralph Kimball? Or is it an outdated rule that can be ignored because of new tech solutions (DW's cheap storage and compute, LookML, other?) If it is a timeless principle, why? And finally, what are some potential solutions?

I've been trying to communicate why we should avoid joins between fact tables in my previous and current job. Find below my take, and I would like to hear yours!

THE PROBLEM
Cross-process analysis requires queries to access multiple fact tables. Because two fact tables share common dimensions, for example client_id, customer_id, or product_id, it appears straightforward to join them using their common surrogate keys. **The problem is that queries that attempt to directly join fact tables using single SQL select clauses can overstate the facts!** Joining across M:M relationship causes over-counting because SQL joins first, then aggregates the “too many rows” create by the join. The only totally safe join between fact tables is when there is a 1:1 relationship, which is very rare and hard to guarantee.The M:M join leads to fact tables without a clear defined grain, with semi-additive and non-additive facts that maybe cannot be summed. This over-counting is known as the “many to many problem”, “fan trap” or “chasm trap”. The derived output fact tables then create a bad UX for analysts and business users. In addition, **performance is poor when millions of facts are joined.**

SOLUTION - DRILL ACROSS QUERIES

BI applications can avoid the M:M problem by performing drill-across queries with multi-pass SQL. Multi pass SQL summarizes fact tables one at a time, then joins the results at a summary level, for example on a customer dimension table. This helps manage performance by keeping individual queries simple and works very well for summary-level process comparisons.

COMPLICATION

However, drill across doesn’t work for every type of cross-process or multi-event analysis. For example: “What was the avg duration between an email being delivered and an email being opened?". The problem is these questions require individual line items from each transaction fact table to be compared before they are aggregated, aka “atomic-level comparisons”. This can result in multi-pass SQL that returns millions of rows that the BI tools or DW must attempt to merge. Even with the correct joins and logic, performance can still be poor.

SOLUTION - DERIVED FACT TABLE PATTERNS

The above comparison problem are caused by poor or missing DW design. Specific email activities are not discrete events that can be fully analyzed in isolation using transaction fact tables. They are evolving event milestones that constantly need to be compared to each other.

“Missing evolving events cause BI pain and suffering” - _Lawrence Corr with Jim Stagnitto_ **Agile Data Warehouse Design** 

Derived fact tables are built from existing fact tables to simplify queries. They use additional ETL processing or additional T layer, and DBMS storage, rather than more complex BI and SQL, to answer difficult analytical questions. Queries can be simplified by derived fact tables and other ETL enhancements,  and **good design will improve performance.**

Definitions:
Event granularity: The combination of event details that guarantee row uniqueness in an event or fact table.
Fully Additive (FA) facts: Produce meaningful results when summed using any combination of the available dimensions.
Non-Additive (NA) facts: Cannot be summed, even if they are the same unit of measure.
Semi-Additive (SA) facts: Can be summed up some of the time, but you can't sum it up all of the time.
Discrete events: Are "point-in-time" or short (duration) stories.
Evolving events: Are longer-running stories that often represent a series of discrete events that BI stakeholders view as milestones of a complex/time -consuming business process.
Recurring events: Are periodic measurement stories that occur at predictable intervals (For example: daily, weekly or monthly)

Resources:
Agile Data Warehouse Design by Lawrence Corr with Jim Stagnitto

**Comment by jasnonaz** [comment]
Diego this is a wonderful writeup! We'd definitely love to work with you to turn this writeup into a post on the developer blog if you're interested.

**Reply by diegogalanhoffman** [reply]
Let's do it! What are next steps?

**Comment by johnblust** [comment]
Hey @diegogalanhoffman, really sorry this fell through the cracks! 

We've been working internally on our processes to make Developer Blog content contribution a better experience for the community. As a result, we consolidated everything onto the [docs.getdbt.com repo](https://github.com/dbt-labs/docs.getdbt.com) & set up some Discussion templates to help get you the help you're looking for as quickly as possible & to refine your ideas! 

You put a lot of effort into this discussion already, and we'd love to see it on the new repo. If you're still interested in writing about this, [create a new discussion](https://github.com/dbt-labs/docs.getdbt.com/discussions) and use this content here to answer the questions in the template. From there, the team will be able to guide you through next steps. Let me or @gwenwindflower know if you have any questions about the new Discussions templates!

**Reply by johnblust** [reply]
@diegogalanhoffman Hey, wanted to check in one last time before I mark this as closed in Asana. Would you still be interested in writing this about this in the new repo? Also if you're seeing this in the future, feel free to add as a discussion in the new repo whenever you have time!

**Reply by diegogalanhoffman** [reply]
Hey @johnblust,

I would be interested in writing this blog post idea in the new repo and following the new process. I'll add it to my action items for the week.

Best,
Diego
